src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs

```diff
@@ -161,31 +175,101 @@ protected void Clone(string repoUri, string commit, string targetDirectory, ILog
-                                Password = pat
-                            }
-                        });
+                        cloneOptions);
 
                     _logger.LogDebug($"Reading local repo from {repoPath}");
                     using (LibGit2Sharp.Repository localRepo = new LibGit2Sharp.Repository(repoPath))
                     {
-                        _logger.LogDebug($"Checking out {commit} in {repoPath}");
-                        LibGit2Sharp.Commands.Checkout(localRepo, commit);
+                        if (commit == null)
+                        {
+                            commit = localRepo.Head.Reference.TargetIdentifier;
+                            _logger.LogInformation($"Repo {localRepo.Info.WorkingDirectory} has no commit to clone at, assuming it's {commit}");
+                        }
+                        try
+                        {
+                            _logger.LogDebug($"Attempting to checkout {commit} as commit in {localRepo.Info.WorkingDirectory}");
+                            LibGit2Sharp.Commands.Checkout(localRepo, commit);
+                        }
+                        catch
+                        {
+                            _logger.LogDebug($"Failed to checkout {commit} as commit, trying to resolve");
+                            string resolvedReference = ParseReference(localRepo, commit, _logger);
+                            _logger.LogDebug($"Resolved {commit} to {resolvedReference ?? "<invalid>"} in {localRepo.Info.WorkingDirectory}, attempting checkout");
+                            LibGit2Sharp.Commands.Checkout(localRepo, resolvedReference);
+                        }
+                    }
+                    // LibGit2Sharp doesn't support a --git-dir equivalent yet (https://github.com/libgit2/libgit2sharp/issues/1467), so we do this manually
+                    if (gitDirectory != null)
+                    {
+                        Directory.Move(repoPath, gitDirectory);
+                        File.WriteAllText(repoPath.TrimEnd('\\', '/'), $"gitdir: {gitDirectory}");
+                    }
+                    using (LibGit2Sharp.Repository localRepo = new LibGit2Sharp.Repository(repoPath))
+                    {
+                        CheckoutSubmodules(localRepo, cloneOptions, gitDirectory, _logger);
                     }
                 }
                 catch (Exception exc)
                 {
-                    throw new Exception($"Something went wrong when cloning repo {repoUri} at {commit} into {targetDirectory}", exc);
+                    throw new Exception($"Something went wrong when cloning repo {repoUri} at {commit ?? "<default branch>"} into {targetDirectory}", exc);
+                }
+            }
+        }
+
+        private static void CheckoutSubmodules(LibGit2Sharp.Repository repo, LibGit2Sharp.CloneOptions submoduleCloneOptions, string gitDirParentPath, ILogger log)
```

Comment:
>private static  [](start = 8, length = 15)

Nit: maybe insert some blank lines in between

Reply:
Spaced out a bit.

---
In reply to: [275519407](https://github.com/dotnet/arcade-services/pull/326#discussion_r275519407) [](ancestors = 275519407)